pkg/apiserver/handlers.go

```diff
@@ -13,9 +20,71 @@ func (s *Server) StartJob(ctx context.Context, req *pb.StartJobRequest) (*pb.Sta
 		// NOTE: Not supposed to happen as already checked, but check anyway.
 		return nil, fmt.Errorf("getUserFromContext: %w", err)
 	}
+	// Contextcheck // False positive. We don't want to use the request context to start the job in the background.
 	jobID, err := s.jobmanager.StartJob(user, req.GetCommand(), req.GetArgs())
 	if err != nil {
-		return nil, fmt.Errorf("jobmanager start job: : %w", err)
+		return nil, fmt.Errorf("job manager start job: : %w", err)
 	}
 	return &pb.StartJobResponse{JobId: jobID.String()}, nil
 }
+
+func (s *Server) StopJob(_ context.Context, req *pb.StopJobRequest) (*pb.StopJobResponse, error) {
+	jobID, err := uuid.Parse(req.GetJobId())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid job id: %s", err)
+	}
+
+	if err := s.jobmanager.StopJob(jobID); err != nil {
+		return nil, fmt.Errorf("job manager stop job: %w", err)
+	}
+
+	return &pb.StopJobResponse{}, nil
+}
+
+func (s *Server) GetJobStatus(_ context.Context, req *pb.GetJobStatusRequest) (*pb.GetJobStatusResponse, error) {
+	jobID, err := uuid.Parse(req.GetJobId())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid job id: %s", err)
+	}
+
+	job, err := s.jobmanager.LookupJob(jobID)
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "lookup job: %s", err)
+	}
+	resp := &pb.GetJobStatusResponse{Status: job.Status()}
+	if resp.GetStatus() != pb.JobStatus_JOB_STATUS_RUNNING {
+		//nolint:gosec // False positive about int/int32 conversion, but in POSIX, exit codes are actually uint8.
+		exitCode := int32(job.ExitCode())
+		resp.ExitCode = &exitCode
+	}
+	return resp, nil
+}
+
+func (s *Server) StreamLogs(req *pb.StreamLogsRequest, ss grpc.ServerStreamingServer[pb.StreamLogsResponse]) error {
+	ctx := ss.Context()
+	jobID, err := uuid.Parse(req.GetJobId())
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "invalid job id: %s", err)
+	}
+
+	r, err := s.jobmanager.StreamLogs(ctx, jobID)
+	if err != nil {
+		return fmt.Errorf("job manager stream logs: %w", err)
+	}
+
+	buf := make([]byte, 32*1024) //nolint:mnd // Default value from io.Copy, reasonable.
+	for {
+		n, err := r.Read(buf)
+		if n > 0 {
+			if err := ss.Send(&pb.StreamLogsResponse{Data: buf[:n]}); err != nil {
+				return fmt.Errorf("send log entry: %w", err)
+			}
+		}
+		if err != nil {
+			if errors.Is(err, io.EOF) || errors.Is(err, io.ErrClosedPipe) {
```

Comment:
The io.EOF handling I get, but why io.ErrClosedPipe?

Reply:
Added comment. When the process dies while in StreamLogs(), we can get a closed pipe error. It doesn't impact the data as the historical output gets locked.